libpod/container_copy_common.go

```diff
@@ -51,6 +51,29 @@ func (c *Container) copyFromArchive(path string, chown, noOverwriteDirNonDir boo
 		}
 	}
 
+	// Before we proceed, mount all named volumes associated with the
+	// container.
+	// This solves two issues:
+	// Firstly, it ensures that if the volume actually requires a mount, we
+	// will mount it for safe use.
+	// (For example, image volumes, volume plugins).
+	// Secondly, it copies up into the volume if necessary.
+	// This ensures that permissions are correct for copies into volumes on
+	// containers that have never started.
+	for _, v := range c.config.NamedVolumes {
+		vol, err := c.mountNamedVolume(v, mountPoint)
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			vol.lock.Lock()
+			if err := vol.unmount(false); err != nil {
+				logrus.Errorf("Unmounting volume %s after container %s copy: %v", vol.Name(), c.ID(), err)
+			}
+			vol.lock.Unlock()
+		}()
```

Comment:
This does not look right, this function here isn't doing any work it is returning another function so defer runs before the copy, you must add that to `unmount()` in the if else case above.